File: chapters/08_queue_and_concurrency_lanes/v2_background_queue.ts
// TypeScript mirror for Chapter 08: Queue and Concurrency Lanes (v2)
// Run with: npx ts-node v2_background_queue.ts

class BackgroundQueue {
  pending: string[] = [];
  done: string[] = [];

  enqueue(taskName: string): void {
    this.pending.push(taskName);
  }

  processNext(): string | null {
    if (this.pending.length === 0) {
      return null;
    }
    const task = this.pending.shift()!;
    this.done.push(task);
    return task;
  }

  progress(): { pending: number; done: number } {
    return {
      pending: this.pending.length,
      done: this.done.length,
    };
  }
}

function main(): void {
  const q = new BackgroundQueue();
  q.enqueue("task-1");
  q.enqueue("task-2");
  q.processNext();
  console.log(q.progress());
}

main();

File: chapters/08_queue_and_concurrency_lanes/v2_background_queue.py
from __future__ import annotations


class BackgroundQueue:
    def __init__(self) -> None:
        self.pending: list[str] = []
        self.done: list[str] = []

    def enqueue(self, task_name: str) -> None:
        self.pending.append(task_name)

    def process_next(self) -> str | None:
        if not self.pending:
            return None
        task = self.pending.pop(0)
        self.done.append(task)
        return task

    def progress(self) -> dict[str, int]:
        return {
            "pending": len(self.pending),
            "done": len(self.done),
        }


def main() -> None:
    q = BackgroundQueue()
    q.enqueue("task-1")
    q.enqueue("task-2")
    q.process_next()
    print(q.progress())


if __name__ == "__main__":
    main()
